modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/cluster/programs/Joining.scala

```diff
@@ -190,6 +77,50 @@ sealed abstract class Joining[F[_]: Async: GenUUID: SecurityProvider: KryoSerial
     } yield ()
   }.onError(err => logger.error(err)(s"Error during join attempt by ${joinRequest.registrationRequest.id.show}"))
 
+  private def validateJoinConditions(): F[Unit] =
+    for {
+      nodeState <- nodeStorage.getNodeState
+      canJoinCluster <- nodeStorage.canJoinCluster
+      _ <- Applicative[F].unlessA(canJoinCluster)(NodeStateDoesNotAllowForJoining(nodeState).raiseError[F, Unit])
+    } yield ()
+
+  private def joinAndDiscover(context: P2PContext): EitherT[F, Throwable, Set[P2PContext]] =
+    EitherT(twoWayHandshake(context, none).attempt).flatMapF { peer =>
+      clusterStorage.getPeer(peer.id).map {
+        case Some(p) => p.asRight[Throwable]
+        case None    => new Exception(s"Peer ${peer.show} was not found").asLeft[Peer]
+      }
+    }.flatMapF { peer =>
+      peerDiscovery
+        .discoverFrom(peer)
+        .attempt
+    }
+      .map(_.map(toP2PContext))
+
+  private def joinAll(toJoin: PeerToJoin): F[Unit] = {
+    type Agg = (Set[P2PContext], Set[P2PContext])
+    type Res = Unit
+
+    (Set(toJoin): Set[P2PContext], Set.empty[P2PContext]).tailRecM {
+      case (joining, _) if joining.isEmpty =>
+        ().asRight[Agg].pure
+      case (joining, attempted) =>
+        for {
+          discovered <- joining.toList.parTraverse { peer =>
+            joinAndDiscover(peer).leftMap { err =>
+              logger.warn(err)(s"Joining to and discovering from peer ${peer.show} failed")
```

Comment:
Seems like this log may never happen given that you are mapping rather than flatMapping (it will be forever delayed I think)

Reply:
Changed to

```scala
        for {
          discovered <- toJoin.toList.parTraverse { peer =>
            joinAndDiscover(peer).valueOrF { err =>
              logger
                .warn(err)(s"Failed to join and discover from ${peer.show}")
                .as(Set.empty[P2PContext])
            }
          }
          reduced = discovered.combineAll
          updatedAttempted = attempted ++ toJoin
          updatedToJoin = reduced.diff(updatedAttempted)
        } yield (updatedToJoin, updatedAttempted).asLeft[Res]
```